documentation/Calls.md

```diff
@@ -0,0 +1,636 @@
+compiler knows that the effective argument data are contiguous
+when compiling the caller (a temporary is needed for other reasons,
+or the effective argument is simply contiguous) and a run-time test for
+discontiguity in the callee can be avoided by using a caller-compaction
+convention when we have the freedom to choose.
+
+While we are unlikely to want to _needlessly_ use a temporary for
+an effective argument that does not require one for any of these
+reasons above, we are specifically disallowed from doing so
+by the standard in cases where pointers to the original target
+data are required to be valid across the call (15.5.2.4(9-10)).
+In particular, compaction of assumed-shape arrays for discretionary
+contiguity on the leading dimension to ease SIMD vectorization
+cannot be done safely for `TARGET` dummies without `VALUE`.
+
+Effective arguments associated with known `INTENT(OUT)` dummies that
+require allocation of a temporary -- and this can only be for reasons of
+contiguity -- don't have to populate it, but they do have to perform
+minimal initialization of any `ALLOCATABLE` components so that
+the runtime doesn't crash when the callee finalizes and deallocates
+them.
+Note that calls to implicit interfaces must conservatively allow
+for the use of `INTENT(OUT)` by the callee.
+
+Except for `VALUE` and known `INTENT(IN)` dummy arguments, the original
+contents of local designators that have been compacted into temporaries
+could optionally have their `ALLOCATABLE` components invalidated
+across the call as an aid to debugging.
+
+Except for `VALUE` and known `INTENT(IN)` dummy arguments, the contents of
+the temporary storage will be copied back into the effective argument
+designator after control returns from the procedure, and it may be necessary
+to preserve addresses (or the values of subscripts and cosubscripts
+needed to recalculate them) of the effective argument designator, or its
+elements, in additional temporary storage if they can't be safely or
+quickly recomputed after the call.
+
+### `INTENT(OUT)` preparation
+
+Effective arguments that are associated with `INTENT(OUT)`
+dummy arguments are required to be definable.
+This cannot always be checked, as the use of `INTENT(OUT)`
+does not by itself mandate the use of an explicit interface.
+
+`INTENT(OUT)` arguments are finalized (as if) on entry to the called
+procedure.  In particular, in calls to elemental procedures,
+the elements of an array are finalized by a scalar or elemental
+`FINAL` procedure (7.5.6.3(7)).
+
+Derived type components that are `ALLOCATABLE` are finalized
+and deallocated.
+Components with initializers are (re)initialized.
+
+The preparation of effective arguments for `INTENT(OUT)` could be
+done on either side of the call.  If the preparation is
+done by the caller, there is an optimization opportunity
+in situations where unmodified incoming `INTENT(OUT)` dummy
+arguments are being passed onward as outgoing `INTENT(OUT)`
+arguments.
```

Comment:
Would it be needed to prove that finalization of the argument has no side effects to do such optimization ?